File: backend/app/schemas/anomaly.py
"""
backend/app/schemas/anomaly.py
Pydantic schemas for automated anomaly and outlier detection engine.
"""
from typing import List, Dict, Any, Optional
from pydantic import BaseModel


class AnomalyPoint(BaseModel):
    row_index: int
    column_name: str
    value: float
    mean: float
    std_dev: float
    z_score: float
    severity: str                           # "EXTREME" (|Z| > 3.5), "MODERATE" (|Z| > 2.5), "MILD" (|Z| > 2.0)
    row_data: Dict[str, Any]


class AnomalySummary(BaseModel):
    total_anomalies: int
    extreme_count: int
    moderate_count: int
    mild_count: int
    anomaly_rate_pct: float
    most_anomalous_column: Optional[str] = None


class AnomalyResponse(BaseModel):
    dataset_id: str
    dataset_name: str
    analyzed_columns: List[str]
    threshold_z: float
    anomalies: List[AnomalyPoint]
    summary: AnomalySummary

File: frontend/types/anomaly.ts
// frontend/types/anomaly.ts

export interface AnomalyPoint {
  row_index: number;
  column_name: string;
  value: number;
  mean: number;
  std_dev: number;
  z_score: number;
  severity: "EXTREME" | "MODERATE" | "MILD";
  row_data: Record<string, unknown>;
}

export interface AnomalySummary {
  total_anomalies: number;
  extreme_count: number;
  moderate_count: number;
  mild_count: number;
  anomaly_rate_pct: number;
  most_anomalous_column?: string | null;
}

export interface AnomalyResponse {
  dataset_id: string;
  dataset_name: string;
  analyzed_columns: string[];
  threshold_z: number;
  anomalies: AnomalyPoint[];
  summary: AnomalySummary;
}
